source/agora/consensus/protocol/Nominator.d

```diff
@@ -526,8 +580,114 @@ extern(D):
 
     public override void signEnvelope (ref SCPEnvelope envelope)
     {
-        envelope.signature = sign(this.schnorr_pair,
-            SCPStatementHash(&envelope.statement).hashFull());
+        // if we're ready to confirm then derive the block and sign its hash
+        if (envelope.statement.pledges.type_ == SCPStatementType.SCP_ST_CONFIRM)
+            this.signConfirmBallot(envelope);
+
+        const Scalar challenge = SCPStatementHash(&envelope.statement).hashFull();
+        envelope.signature = sign(this.schnorr_pair, challenge);
+        log.trace("SIGN Envelope signature {} \nfor public key {} \n" ~
+                "envelope {}\n challenge {}",
+                envelope.signature, PublicKey(this.schnorr_pair.V[]),
+                scpPrettify(&envelope), challenge);
+    }
+
+    /***************************************************************************
+
+        Signs the confirm ballot in the SCPEnvelope
+
+        Params:
+            envelope = the SCPEnvelope
+
+    ***************************************************************************/
+
+    private void signConfirmBallot (ref SCPEnvelope envelope) nothrow
+    {
+        assert(envelope.statement.pledges.type_ ==
+            SCPStatementType.SCP_ST_CONFIRM);
+
+        ConsensusData con_data;
+        try
+        {
+            con_data = deserializeFull!ConsensusData(
+                envelope.statement.pledges.confirm_.ballot.value[]);
+        }
+        catch (Exception ex)
+        {
+            assert(0);  // this should never happen
+        }
+
+        auto blocks = this.ledger.getBlocksFrom(
+            Height(envelope.statement.slotIndex - 1));
+        assert(!blocks.empty);
+
+        const prev_block = blocks.front;
+        const proposed_block = makeNewBlock(prev_block, con_data.tx_set,
+            con_data.enrolls);
+
+        Height height = proposed_block.header.height;
+        const PublicKey public_key = PublicKey(this.schnorr_pair.V[]);
+
+        // challenge = Hash(block) to Scalar
+        const Scalar challenge = hashFull(proposed_block);
+
+        // rc = r used in signing the commitment
+        const Scalar rc = this.enroll_man.getCommitmentNonceScalar();
+        const Scalar r = rc + challenge; // make it unique each challenge
+        const Pair R = Pair.fromScalar(r);
+
+        const Signature signature = multiSigSign(R, this.schnorr_pair.v, challenge);
+        envelope.statement.pledges.confirm_.value_sig = opaque_array!64(BitBlob!512(signature[]));
+        auto block_sig = ValidatorBlockSig(height, public_key, signature);
+
+        // Store our block signature in the slot_sigs map
+        this.slot_sigs[block_sig.height][this.schnorr_pair.V] = signature;
+        this.network.gossipBlockSignature(block_sig);
```

Comment:
I think it might be a little bit premature to publish it at this point. Other nodes may not have reached the confirm stage yet, in which case they may decide to reject your signature which was just gossiped. How about only publishing our own signature after `externalize` instead?

Reply:
Moved to after accepting new block.